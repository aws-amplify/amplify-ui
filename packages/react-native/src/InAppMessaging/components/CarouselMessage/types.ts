import { StyleProp, ViewStyle } from 'react-native';

import {
  CarouselMessageCommonProps,
  MessageComponentBaseProps,
} from '@aws-amplify/ui-react-core-notifications';

import { MessageDefaultStyle, MessageOverrideStyle } from '../../hooks';

interface CarouselMessageOverrideStyle extends MessageOverrideStyle {
  pageIndicatorActive?: StyleProp<ViewStyle>;
  pageIndicatorInactive?: StyleProp<ViewStyle>;
}

export interface CarouselMessageProps
  extends CarouselMessageCommonProps<
    CarouselMessageOverrideStyle | undefined
  > {}

export interface CarouselMessageItemProps
  extends MessageComponentBaseProps<CarouselMessageOverrideStyle> {}

export interface CarouselMessageStyle extends MessageDefaultStyle {}

export interface CarouselMessageComponentStyle {
  pageIndicatorActive: ViewStyle;
  pageIndicatorInactive: ViewStyle;
}
