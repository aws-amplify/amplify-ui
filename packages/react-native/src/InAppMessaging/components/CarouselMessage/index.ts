export { default as CarouselMessage } from './CarouselMessage';
export { CarouselMessageProps } from './types';
