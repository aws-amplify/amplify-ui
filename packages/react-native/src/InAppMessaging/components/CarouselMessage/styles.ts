import { ImageStyle, StyleSheet } from 'react-native';
import {
  BORDER_RADIUS_BASE,
  COLOR_GREY,
  COLOR_LIGHT_GREY,
  COLOR_LIGHT_GREY_2,
  COLOR_WHITE,
  FONT_SIZE_BASE,
  FONT_SIZE_LARGE,
  FONT_WEIGHT_BASE,
  FONT_WEIGHT_BOLD,
  LINE_HEIGHT_BASE,
  LINE_HEIGHT_LARGE,
  SPACING_EXTRA_LARGE,
  SPACING_LARGE,
  SPACING_MEDIUM,
  SPACING_SMALL,
} from '../../constants';

import { CarouselMessageStyle, CarouselMessageComponentStyle } from './types';

export const defaultStyle: CarouselMessageComponentStyle = StyleSheet.create({
  pageIndicatorActive: {
    backgroundColor: COLOR_GREY,
    borderRadius: SPACING_LARGE / 2,
    height: SPACING_LARGE,
    margin: SPACING_SMALL,
    width: SPACING_LARGE,
  },
  pageIndicatorInactive: {
    backgroundColor: COLOR_LIGHT_GREY_2,
    borderRadius: SPACING_LARGE / 2,
    height: SPACING_LARGE,
    margin: SPACING_SMALL,
    width: SPACING_LARGE,
  },
});

const commonStyles: Omit<CarouselMessageStyle, 'image'> = {
  body: {
    fontSize: FONT_SIZE_BASE,
    fontWeight: FONT_WEIGHT_BASE,
    lineHeight: LINE_HEIGHT_BASE,
  },
  buttonContainer: {
    backgroundColor: COLOR_LIGHT_GREY,
    borderRadius: BORDER_RADIUS_BASE,
    flex: 1,
    margin: SPACING_MEDIUM,
    padding: SPACING_LARGE,
  },
  buttonsContainer: {
    flexDirection: 'row',
    justifyContent: 'center',
  },
  buttonText: {
    fontSize: FONT_SIZE_BASE,
    fontWeight: FONT_WEIGHT_BASE,
    lineHeight: LINE_HEIGHT_BASE,
    textAlign: 'center',
  },
  container: {
    flex: 1,
    padding: SPACING_EXTRA_LARGE,
  },
  contentContainer: {
    flex: 1,
  },
  header: {
    fontSize: FONT_SIZE_LARGE,
    fontWeight: FONT_WEIGHT_BOLD,
    lineHeight: LINE_HEIGHT_LARGE,
  },
  iconButton: {
    alignSelf: 'flex-start',
    marginBottom: SPACING_MEDIUM,
    marginLeft: 'auto',
    marginRight: SPACING_MEDIUM,
  },
  imageContainer: {
    alignItems: 'center',
    margin: SPACING_LARGE,
  },
  textContainer: {
    paddingHorizontal: SPACING_MEDIUM,
  },
  wrapper: {
    backgroundColor: COLOR_WHITE,
    flex: 1,
  },
};

export const getPortraitStyles = (
  imageDimensions: ImageStyle
): CarouselMessageStyle =>
  StyleSheet.create({
    ...commonStyles,
    image: { ...imageDimensions },
    textContainer: {
      ...commonStyles.textContainer,
      marginVertical: SPACING_LARGE,
    },
  });

export const getLandscapeStyles = (
  imageDimensions: ImageStyle
): CarouselMessageStyle =>
  StyleSheet.create({
    ...commonStyles,
    contentContainer: {
      ...commonStyles.contentContainer,
      alignContent: 'center',
      justifyContent: 'center',
      flexDirection: 'row',
    },
    image: { ...imageDimensions },
    imageContainer: {
      ...commonStyles.imageContainer,
      justifyContent: 'center',
    },
    textContainer: {
      ...commonStyles.textContainer,
      flex: 1,
      justifyContent: 'center',
    },
  });
