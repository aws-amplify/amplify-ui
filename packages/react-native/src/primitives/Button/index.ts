export { default as Button } from './Button';
export { ButtonProps, ButtonStyles } from './types';
