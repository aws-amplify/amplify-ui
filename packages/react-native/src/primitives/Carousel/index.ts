export { default as Carousel } from './Carousel';
export { DEFAULT_CAROUSEL_INDICATOR_SIZE } from './constants';
export { CarouselProps } from './types';
