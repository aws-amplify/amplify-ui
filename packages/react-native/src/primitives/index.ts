export * from './Button';
export * from './Carousel';
export * from './Checkbox';
export * from './Divider';
export * from './ErrorMessage';
export * from './Heading';
export * from './Icon';
export * from './IconButton';
export * from './Label';
export * from './PasswordField';
export * from './PhoneNumberField';
export * from './Radio';
export * from './RadioGroup';
export * from './Tabs';
export * from './TextField';
