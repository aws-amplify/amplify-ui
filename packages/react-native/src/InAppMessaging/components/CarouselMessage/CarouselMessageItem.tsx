import React from 'react';
import { SafeAreaView } from 'react-native-safe-area-context';

import { useDeviceOrientation } from '../../../hooks';
import { useMessageProps } from '../../hooks';
import { MessageLayout } from '../MessageLayout';

import { getLandscapeStyles, getPortraitStyles } from './styles';
import { CarouselMessageItemProps } from './types';

export default function CarouselMessageItem(
  props: CarouselMessageItemProps
): JSX.Element | null {
  const { deviceOrientation, isPortraitMode } = useDeviceOrientation();
  const messageProps = useMessageProps(
    props,
    isPortraitMode ? getPortraitStyles : getLandscapeStyles
  );
  const { shouldRenderMessage, styles } = messageProps;

  if (!shouldRenderMessage) {
    return null;
  }

  const { wrapper, ...messageStyles } = styles!;

  return (
    <SafeAreaView style={wrapper}>
      <MessageLayout
        {...props}
        {...messageProps}
        orientation={deviceOrientation}
        styles={messageStyles}
      />
    </SafeAreaView>
  );
}
