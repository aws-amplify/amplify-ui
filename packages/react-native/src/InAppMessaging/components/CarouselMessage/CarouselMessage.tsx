import React, { useMemo } from 'react';
import { MessageContentProps } from '@aws-amplify/ui-react-core-notifications';

import { Carousel } from '../../../primitives';
import { IN_APP_MESSAGING_TEST_ID } from '../../constants';

import { MessageWrapper } from '../MessageWrapper';
import CarouselMessageItem from './CarouselMessageItem';
import { defaultStyle } from './styles';
import { CarouselMessageProps } from './types';

export default function CarouselMessage(
  props: CarouselMessageProps
): JSX.Element {
  const { data, ...rest } = props;
  const { style } = rest;

  const indicatorStyle = useMemo(
    () => ({
      active: [defaultStyle.pageIndicatorActive, style?.pageIndicatorActive],
      inactive: [
        defaultStyle.pageIndicatorInactive,
        style?.pageIndicatorInactive,
      ],
    }),
    [style]
  );

  const renderItem = ({ item }: { item: MessageContentProps }) => (
    <CarouselMessageItem {...item} {...rest} />
  );

  return (
    <MessageWrapper disableSafeAreaView>
      <Carousel
        data={data ?? []}
        renderItem={renderItem}
        indicatorActiveStyle={indicatorStyle.active}
        indicatorInactiveStyle={indicatorStyle.inactive}
        testID={IN_APP_MESSAGING_TEST_ID.CAROUSEL}
      />
    </MessageWrapper>
  );
}
