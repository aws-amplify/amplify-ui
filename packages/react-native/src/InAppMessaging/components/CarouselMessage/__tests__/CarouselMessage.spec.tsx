import React, { ReactElement } from 'react';
import { act, create, ReactTestRenderer } from 'react-test-renderer';

import { Carousel } from '../../../../primitives';
import CarouselMessage from '../CarouselMessage';

jest.mock('../../../../primitives', () => ({ Carousel: 'Carousel' }));
jest.mock('../../MessageWrapper', () => ({ MessageWrapper: 'MessageWrapper' }));
jest.mock('../CarouselMessageItem', () => 'CarouselMessageItem');

const baseProps = { layout: 'CAROUSEL' as const, data: [] };

describe('CarouselMessage', () => {
  let carouselMessage: ReactTestRenderer;
  let itemRenderer: ReactTestRenderer;

  it('renders as expected', () => {
    act(() => {
      carouselMessage = create(<CarouselMessage {...baseProps} />);
    });

    expect(carouselMessage.toJSON()).toMatchSnapshot();
  });

  it('allows style overrides', () => {
    const overrides = {
      pageIndicatorActive: { backgroundColor: 'red' },
      pageIndicatorInactive: { backgroundColor: 'blue' },
    };
    act(() => {
      carouselMessage = create(
        <CarouselMessage {...baseProps} style={overrides} />
      );
    });

    expect(carouselMessage.toJSON()).toMatchSnapshot();
  });

  it('renders items', () => {
    act(() => {
      carouselMessage = create(<CarouselMessage {...baseProps} />);
    });
    const carousel = carouselMessage.root.findByType(Carousel);
    const Item = () =>
      (carousel.props as { renderItem: (data) => ReactElement }).renderItem({
        item: { image: { src: 'image-src' } },
      });

    act(() => {
      itemRenderer = create(<Item />);
    });

    expect(itemRenderer.toJSON()).toMatchSnapshot();
  });
});
