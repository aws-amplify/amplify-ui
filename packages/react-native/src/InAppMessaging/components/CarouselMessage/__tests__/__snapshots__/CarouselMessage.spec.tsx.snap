// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CarouselMessage allows style overrides 1`] = `
<MessageWrapper
  disableSafeAreaView={true}
>
  <Carousel
    data={Array []}
    indicatorActiveStyle={
      Array [
        Object {
          "backgroundColor": "#a1a1a1",
          "borderRadius": 6,
          "height": 12,
          "margin": 4,
          "width": 12,
        },
        Object {
          "backgroundColor": "red",
        },
      ]
    }
    indicatorInactiveStyle={
      Array [
        Object {
          "backgroundColor": "#d8d8d8",
          "borderRadius": 6,
          "height": 12,
          "margin": 4,
          "width": 12,
        },
        Object {
          "backgroundColor": "blue",
        },
      ]
    }
    renderItem={[Function]}
  />
</MessageWrapper>
`;

exports[`CarouselMessage renders as expected 1`] = `
<MessageWrapper
  disableSafeAreaView={true}
>
  <Carousel
    data={Array []}
    indicatorActiveStyle={
      Array [
        Object {
          "backgroundColor": "#a1a1a1",
          "borderRadius": 6,
          "height": 12,
          "margin": 4,
          "width": 12,
        },
        undefined,
      ]
    }
    indicatorInactiveStyle={
      Array [
        Object {
          "backgroundColor": "#d8d8d8",
          "borderRadius": 6,
          "height": 12,
          "margin": 4,
          "width": 12,
        },
        undefined,
      ]
    }
    renderItem={[Function]}
  />
</MessageWrapper>
`;

exports[`CarouselMessage renders items 1`] = `
<CarouselMessageItem
  image={
    Object {
      "src": "image-src",
    }
  }
  layout="CAROUSEL"
/>
`;
